vendor_blueprints/export_card_to_file.py

```diff
@@ -0,0 +1,166 @@
+        "method": "password",
+        "emailAddress": email,
+        "password": password
+    })
+
+    auth_headers = {'Content-Type' : 'application/json'}
+    try:
+        auth_response = requests.post(auth_api, data=auth_body, 
+                                      headers=auth_headers)
+    except Exception as e:
+        print(f"Request error: {e}")
+        sys.exit(EXIT_CODE_BAD_REQUEST)
+
+    auth_response_json = auth_response.json()
+    if auth_response_json["success"] is False: # Failed to login
+        print(f"Authentication failed due to reason: {auth_response_json['reason']}")
+        sys.exit(EXIT_CODE_INVALID_CREDENTIALS)
+        
+    # else if the authentication succeeded
+    domo_token = auth_response_json['sessionToken']
+    return domo_token
+
+
+def get_card_data(card_id, access_token, DOMO_INSTANCE):
+    """
+    Get metadata and property information of a single card
+
+    card_id (str): The unique ID of the card
+    access_token: client access token. Use the get_access_token() function to retrieve the token.
+
+    Returns:
+    card_response -> dict with the metadata and details of the card.
+    """
+    card_info_api = f"https://{DOMO_INSTANCE}.domo.com/api/content/v1/cards?urns={card_id}&parts=metadata,properties&includeFiltered=true"
+    card_headers = {
+            'Content-Type' : 'application/json',
+            'x-domo-authentication': access_token
+        }
+    card_response = requests.get(url=card_info_api, headers=card_headers)
+    return card_response.json()
+
+
+def export_graph_to_file(card_id, file_name, file_type, 
+                         access_token, DOMO_INSTANCE, folder_path=""):
+    """
+    Exports a file to one of the given file types: csv, ppt, excel
+    """
+    export_api = f"https://{DOMO_INSTANCE}.domo.com/api/content/v1/cards/{card_id}/export"
+
+    card_headers = {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'accept': 'application/json, text/plain, */*',
+            'x-domo-authentication': access_token
+    }
+    
+    # make a dictionary to map user file_type with requested mimetype
+    filetype_map = {
+        "csv":"text/csv", 
+        "excel":"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+        "ppt": "application/vnd.ms-powerpoint"
+    }
+    
+    body = {
+        "queryOverrides":{
+            "filters":[],
+            "dataControlContext":{
+                "filterGroupIds":[]
+            }
+        },
+        "watermark":'true',
+        "mobile":'false',
+        "showAnnotations":'true',
+        "type":"file",
+        "fileName": f"{file_name}",
+        "accept": filetype_map[file_type]
+    }
+    # convert body to domo encoded payload
+    encoded_body = urllib.parse.quote(f"{body}")
+    encoded_body = encoded_body.replace("%27", "%22") # changing " to '
+    payload = f"request={encoded_body}"
+
+    export_response = requests.post(url=export_api, data=payload,
```

Comment:
What's the end goal for `data` with all of this logic? I'm assuming it's to json encode the body and then set that encoded value on the key `request` in the request body. Is the request body meant to be a form? 
If the goal is JSON, then use a normal, typed dict for body and use the json library to encode it. We shouldn't be using quote replacements for body encodings. 
If the goal isn't JSON, then please explain.

Reply:
The major issue with data is that the cards/card_id/export/ route has a very specific payload it recieves which is a POST with a body with the following rules:

1. The entire body has to be a single string
2. the payload HAS TO start with **request=**.
3. the rest of the payload after the **request=** has to a dictionary with all the special characters (curly braces, quotes, etc.) url encoded.
4. The quotes used in the payload for the dict/json after the request= that are url encoded can only be double quotes (aka ""). Python by default converts all dictionary quotes into single quotes so a conversion to double quotes after it's been stringified is neccessary.
5. the double quotes conversion can only be done *after* all the previous url conversion has happened due to some issue that causes single quotes to remain in the string even after conversion

The API itself stricly enforces that and any other format or way of tramsitting the data returns a 400 error which is why the code is in that format in the first place. It's admittedly extremely strange, but that's the way the API works and after trial and error and extensively checking the network data that's how i figured out the format it uses.

Reply:
Okay, that all makes sense. Can you document that the API has weird requirements and that's why the code is the way it is. Essentially summarize what you've posted here in comments. _(with Axolo)_